keps/sig-node/4818-allow-zero-value-for-sleep-action-of-prestop-hook/README.md

```diff
@@ -0,0 +1,564 @@
+
+## Motivation
+
+Currently, trying to create a container with a PreStop lifecycle hook with sleep of 0 seconds will throw a validation error like so:
+
+```
+Invalid value: 0: must be greater than 0 and less than terminationGracePeriodSeconds (30)
+```
+
+The Sleep action is implemented with the time package from Go’s standard library. The `time.After()` which is used to implement the sleep permits a zero sleep duration. A negative or a zero sleep duration will cause the function to return immediately and function like a no-op.
+
+The implementation in KEP 3960 supports only non-zero values for the sleep duration. It is semantically correct to support a zero value for this field since time.After() also supports zero and negative durations. Negative values as well as zero have the same effect with time.After(), they both return immediately. We don’t need to support negative values since they have the same effect as setting the duration to zero.
+
+A potential use case for this behaviour is when you need a PreStop hook to be defined for the validation of your resource, but don't really need to sleep as part of the PreStop hook. An example of this is described by a user [here](https://github.com/kubernetes/enhancements/issues/3960#issuecomment-2208556397) in the parent KEP. They add a PreStop sleep hook in via an admission webhoook by default if the PreStop is hook is not specified by the user. In order to opt-out from this, a no-op PreStop hook with a duration of zero seconds can be used.
+
+### Goals
+
+- Update the validation for the Sleep action to allow zero as a valid sleep duration.
+- Allow users to set a zero value for the sleep action in PreStop hooks to do a no-op.
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+Introduce a `PodLifecycleSleepActionAllowZero` feature gate which is disabled by default. When the feature gate is enabled, the `validateSleepAction` method would allow values greater than or equal to zero as a valid sleep duration.
+
+
+Since this update to the validation allows previously invalid values, care must be taken to support cluster downgrades safely. To accomplish this, the validation will distinguish between new resources and updates to existing resources:
+
+
+- When the feature gate is disabled:
+  - (a) New resources will no longer allow setting zero as the sleep duration second for the PreStop hook. (no change to current validation)
+  - (b) Existing resources cannot be updated to have a sleep duration of zero seconds
+  - (c) Existing resources with a PreStop sleep duration set to zero will continue to run and use a sleep duration of zero seconds. These resources however cannot be updated without also updating the sleep duration to a non-zero value, since the validation would now fail.
```

Comment:
> These resources however cannot be updated

So status updates will not be allowed? You may want to be more precise on what updates will not be allowed

Reply:
Status updates are allowed. [Ref](https://docs.google.com/document/d/1o01gH2ELPY-kjwgemjg3BbWiykx37ag-4VYwwJ1jBTw/edit?disco=AAABWDp70to). Also [discussion in the PR](https://github.com/kubernetes/kubernetes/pull/127094#discussion_r1764005018). I need to update the KEP with this. Will do it today.

Reply:
yep, as expected. I just pointing out that this statement is quite scary and should be clarified.